tests/integration/security/sds_citadel_flow/sds_citadel_flow_test.go

```diff
@@ -0,0 +1,162 @@
+		return nil
+	}
+	return &testPolicy{
+		t:         t,
+		env:       env,
+		namespace: namespace,
+		fileName:  fileName,
+	}
+}
+
+func checkConnection(conn connection) error {
+	ep := conn.to.EndpointForPort(conn.port)
+	if ep == nil {
+		return fmt.Errorf("cannot get upstream endpoint for connection test %v", conn)
+	}
+
+	results, err := conn.from.Call(ep, apps.AppCallOptions{Protocol: conn.protocol})
+	if conn.expectedSuccess {
+		if err != nil || len(results) == 0 || results[0].Code != "200" {
+			// Addition log for debugging purpose.
+			if err != nil {
+				fmt.Printf("Error: %#v\n", err)
+			} else if len(results) == 0 {
+				fmt.Printf("No result\n")
+			} else {
+				fmt.Printf("Result: %v\n", results[0])
+			}
+			return fmt.Errorf("%s to %s:%d using %s: expected success, actually failed",
+				conn.from.Name(), conn.to.Name(), conn.port, conn.protocol)
+		}
+	} else {
+		if err == nil && len(results) > 0 && results[0].Code == "200" {
+			return fmt.Errorf("%s to %s:%d using %s: expected failed, actually success",
+				conn.from.Name(), conn.to.Name(), conn.port, conn.protocol)
+		}
+	}
+	return nil
+}
+
+func TestSdsCitadelCaFlow(t *testing.T) {
+	ctx := framework.NewContext(t)
+	defer ctx.Done(t)
+	ctx.RequireOrSkip(t, environment.Kube)
+
+	istioCfg, err := istio.DefaultConfig(ctx)
+	if err != nil {
+		t.Fatalf("Fail to get default config: %v", err)
+	}
+	env := ctx.Environment().(*kube.Environment)
+	pilot := pilot.NewOrFail(t, ctx, pilot.Config{})
+	appInst := apps.NewOrFail(ctx, t, apps.Config{Pilot: pilot})
+
+	aApp, _ := appInst.GetAppOrFail("a", t).(apps.KubeApp)
+	bApp, _ := appInst.GetAppOrFail("b", t).(apps.KubeApp)
+
+	connections := []connection{
+		{
+			from:            aApp,
+			to:              bApp,
+			port:            80,
+			protocol:        apps.AppProtocolHTTP,
+			expectedSuccess: true,
+		},
+	}
+
+	namespace := istioCfg.SystemNamespace
+	configFile := "global-mtls.yaml"
+
+	policy := applyPolicyFile(t, env, namespace, configFile)
+	// Sleep 3 seconds for the policy to take effect.
+	time.Sleep(3 * time.Second)
+	for _, conn := range connections {
+		retry.UntilSuccessOrFail(t, func() error {
+			return checkConnection(conn)
+		}, retry.Delay(time.Second), retry.Timeout(10*time.Second))
+	}
+	policy.tearDown()
+}
+
+func setupConfig(cfg *istio.Config) {
+	if cfg == nil {
+		return
+	}
+	cfg.Values["sidecarInjectorWebhook.rewriteAppHTTPProbe"] = "true"
+	cfg.Values["global.controlPlaneSecurityEnabled"] = "false"
+	cfg.Values["global.mtls.enabled"] = "true"
+	cfg.Values["global.sds.enabled"] = "true"
+	cfg.Values["global.sds.udsPath"] = "unix:/var/run/sds/uds_path"
+	cfg.Values["global.sds.useNormalJwt"] = "true"
+	cfg.Values["nodeagent.enabled"] = "true"
+	cfg.Values["nodeagent.image"] = "node-agent-k8s"
+	cfg.Values["nodeagent.env.CA_PROVIDER"] = "Citadel"
+	cfg.Values["nodeagent.env.CA_ADDR"] = "istio-citadel:8060"
+	cfg.Values["nodeagent.env.VALID_TOKEN"] = "true"
+}
+
+func TestMain(m *testing.M) {
```

Comment:
FYI you have a couple of tests here. Spinning an entire k8s environment for each will take a while ... is that what you want?

Reply:
The two tests in the PR uses two different CA providers (i.e., Vault CA and Citadel) by running helm template commands with different CA configurations, thereby having two different k8s Istio environments.